<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Drawing LDheatmap from data in VCF format â€¢ LDheatmap</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Drawing LDheatmap from data in VCF format">
<meta property="og:description" content="">
<meta property="og:image" content="https://github.com/mcneney/LDheatmap/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LDheatmap</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fas fa fas fa-paperclip"></span>
     
    Reference
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa fas fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://cran.r-project.org/web/packages/LDheatmap/vignettes/LDheatmap.pdf">Journal</a>
    </li>
    <li>
      <a href="https://cran.r-project.org/web/packages/LDheatmap/vignettes/addTracks.pdf">Example of Adding Tracks</a>
    </li>
    <li>
      <a href="../articles/vcfOnLDheatmap.html">Drawing LDheatmap from data in VCF format</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mcneney/LDheatmap/">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Drawing LDheatmap from data in VCF format</h1>
                        <h4 class="author">Gloria Yang</h4>
            
            <h4 class="date">2018-10-14</h4>
      
      
      <div class="hidden name"><code>vcfOnLDheatmap.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>VCF (Variant Call Format) is a text file format. It contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. There is an example how to do draw LDheatmap from data in VCF format</p>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<p><code>snp_in_vcf.vcf</code> is a vcf datafile contains common SNPs (SNPs with frequency 5% or more in the world-wide population) in the MLLT3 gene. We are going to draw the LDheatmap based on European desent and Asian descent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#read in the vcf data file</span>
<span class="kw">require</span>(vcfR)
<span class="co">#&gt; Loading required package: vcfR</span>
<span class="co">#&gt; Warning: package 'vcfR' was built under R version 3.4.4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    *****       ***   vcfR   ***       *****</span>
<span class="co">#&gt;    This is vcfR 1.8.0 </span>
<span class="co">#&gt;      browseVignettes('vcfR') # Documentation</span>
<span class="co">#&gt;      citation('vcfR') # Citation</span>
<span class="co">#&gt;    *****       *****      *****       *****</span>
snp &lt;-<span class="st"> </span><span class="kw">read.vcfR</span>(<span class="st">"snp_in_vcf.vcf"</span>)
<span class="co">#&gt; Scanning file to determine attributes.</span>
<span class="co">#&gt; File attributes:</span>
<span class="co">#&gt;   meta lines: 252</span>
<span class="co">#&gt;   header_line: 253</span>
<span class="co">#&gt;   variant count: 432</span>
<span class="co">#&gt;   column count: 2513</span>
<span class="co">#&gt; </span>
Meta line <span class="dv">252</span> read in.
<span class="co">#&gt; All meta lines processed.</span>
<span class="co">#&gt; gt matrix initialized.</span>
<span class="co">#&gt; Character matrix gt created.</span>
<span class="co">#&gt;   Character matrix gt rows: 432</span>
<span class="co">#&gt;   Character matrix gt cols: 2513</span>
<span class="co">#&gt;   skip: 0</span>
<span class="co">#&gt;   nrows: 432</span>
<span class="co">#&gt;   row_num: 0</span>
<span class="co">#&gt; </span>
Processed variant<span class="op">:</span><span class="st"> </span><span class="dv">432</span>
<span class="co">#&gt; All variants processed</span></code></pre></div>
<p><code>1KG_sample_info.csv</code> is the data file which contain the information about the sample population and corresponding super population code</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#super population for EUR &amp; EAS</span>
<span class="co">#Get the corresponding population code for EUR &amp; EAS</span>
sample_info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">"1KG_sample_info.csv"</span>)
eur &lt;-<span class="st"> </span>sample_info[sample_info<span class="op">$</span>Population <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"CEU"</span>,<span class="st">"TSI"</span>,<span class="st">"FIN"</span>,<span class="st">"GBR"</span>,<span class="st">"IBS"</span>),<span class="op">-</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]
eas &lt;-<span class="st"> </span>sample_info[sample_info<span class="op">$</span>Population <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"CHB"</span>,<span class="st">"JPT"</span>,<span class="st">"CHS"</span>,<span class="st">"CDX"</span>,<span class="st">"KHV"</span>),<span class="op">-</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]

<span class="co">#all column value from EUR/EAS and first FORMAT column</span>
eur_gt &lt;-<span class="st"> </span>snp<span class="op">@</span>gt[,<span class="kw">colnames</span>(snp<span class="op">@</span>gt) <span class="op">%in%</span><span class="st"> </span>eur[,<span class="dv">1</span>]]
eas_gt &lt;-<span class="st"> </span>snp<span class="op">@</span>gt[,<span class="kw">colnames</span>(snp<span class="op">@</span>gt) <span class="op">%in%</span><span class="st"> </span>eas[,<span class="dv">1</span>]]

eur_snpMat &lt;-<span class="st"> </span><span class="kw">t</span>(eur_gt)
eas_snpMat &lt;-<span class="st"> </span><span class="kw">t</span>(eas_gt)</code></pre></div>
<p>Convert the matrix of genotypes to a numeric matrix in which genotypes are coded as 0, 1 or 2 copies of the minor allele.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#define a function to convert the value into 0,1,2</span>
convertToNumeric &lt;-<span class="st"> </span><span class="cf">function</span>(x){
gdat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow =</span> <span class="kw">nrow</span>(x), <span class="dt">ncol =</span> <span class="kw">ncol</span>(x))
<span class="cf">for</span> (m <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(x)){
    <span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(x)){
      a &lt;-<span class="kw">as.numeric</span>(<span class="kw">unlist</span>(<span class="kw">strsplit</span>(x[m,n], <span class="st">"|"</span>))[<span class="dv">1</span>]) 
      
      b &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">unlist</span>(<span class="kw">strsplit</span>(x[m,n], <span class="st">"|"</span>))[<span class="dv">3</span>])
      gdat[m,n] &lt;-<span class="st"> </span>a<span class="op">+</span>b
    }
}
<span class="kw">rownames</span>(gdat) &lt;-<span class="st"> </span><span class="kw">rownames</span>(x)
<span class="kw">colnames</span>(gdat) &lt;-<span class="st"> </span><span class="kw">colnames</span>(x)
      <span class="kw">return</span>(gdat)
}


<span class="co">#convert to snpMatrix - EUR</span>
gdat_eur &lt;-<span class="st"> </span><span class="kw">convertToNumeric</span>(eur_snpMat)

<span class="co">#load the snp_id_dist.csv, which contains the SNPs id and distance</span>
info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">"snp_id_dist.csv"</span>)
snpNames &lt;-<span class="st"> </span>info<span class="op">$</span>id
<span class="kw">colnames</span>(gdat_eur) &lt;-<span class="st"> </span>snpNames

<span class="kw">require</span>(snpStats)
<span class="co">#&gt; Loading required package: snpStats</span>
<span class="co">#&gt; Loading required package: survival</span>
<span class="co">#&gt; Warning: package 'survival' was built under R version 3.4.4</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Warning: package 'Matrix' was built under R version 3.4.4</span>
<span class="kw">library</span>(LDheatmap)
<span class="co">#&gt; Warning: package 'LDheatmap' was built under R version 3.4.4</span>
gdat_eur&lt;-<span class="kw">as</span>(gdat_eur,<span class="st">"SnpMatrix"</span>)
<span class="co">#&gt; Warning in asMethod(object): values other than 0, 1 or 2 set to NA</span>
<span class="kw"><a href="../reference/LDheatmap.html">LDheatmap</a></span>(gdat_eur,info<span class="op">$</span>filt_snpDist,<span class="dt">title=</span><span class="st">'Europeans'</span>,<span class="dt">add.map=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="vcfOnLDheatmap_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co">#convert to snpMatrix - EAS</span>
gdat_eas &lt;-<span class="st"> </span><span class="kw">convertToNumeric</span>(eas_snpMat)
<span class="kw">colnames</span>(gdat_eas) &lt;-<span class="st"> </span>snpNames
gdat_eas&lt;-<span class="kw">as</span>(gdat_eas,<span class="st">"SnpMatrix"</span>)
<span class="co">#&gt; Warning in asMethod(object): values other than 0, 1 or 2 set to NA</span>
<span class="kw"><a href="../reference/LDheatmap.html">LDheatmap</a></span>(gdat_eas,info<span class="op">$</span>filt_snpDist,<span class="dt">title=</span><span class="st">'Asians'</span>,<span class="dt">add.map=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="vcfOnLDheatmap_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#getting-started">Getting started</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Ji-Hyung Shin, Sigal Blay, Brad McNeney, Jinko Graham.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
