<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>This function produces a pairwise LD plot. — LDheatmap • LDheatmap</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="This function produces a pairwise LD plot. — LDheatmap" />

<meta property="og:description" content="LDheatmap() is used to produce a graphical display, as a heat map,
of pairwise linkage disequilibrium (LD) measurements for SNPs.
The heat map is a false color image in the upper-left diagonal of a square plot.
Optionally, a line parallel to the diagonal of the image indicating
the physical or genetic map positions of the SNPs may be added, along
with text reporting the total length of the genomic region considered." />

<meta property="og:image" content="https://github.com/mcneney/LDheatmap/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LDheatmap</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fas fa fas fa-paperclip"></span>
     
    Reference
  </a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fas fa fas fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://cran.r-project.org/web/packages/LDheatmap/vignettes/LDheatmap.pdf">Journal</a>
    </li>
    <li>
      <a href="https://cran.r-project.org/web/packages/LDheatmap/vignettes/addTracks.pdf">Example of Adding Tracks</a>
    </li>
    <li>
      <a href="../articles/vcfOnLDheatmap.html">Drawing LDheatmap from data in VCF format</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mcneney/LDheatmap/">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>This function produces a pairwise LD plot.</h1>
    
    <div class="hidden name"><code>LDheatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>LDheatmap()</code> is used to produce a graphical display, as a heat map,
of pairwise linkage disequilibrium (LD) measurements for SNPs.
The heat map is a false color image in the upper-left diagonal of a square plot.
Optionally, a line parallel to the diagonal of the image indicating
the physical or genetic map positions of the SNPs may be added, along
with text reporting the total length of the genomic region considered.</p>
    
    </div>

    <pre class="usage"><span class='fu'>LDheatmap</span>(<span class='no'>gdat</span>, <span class='kw'>genetic.distances</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>distances</span><span class='kw'>=</span><span class='st'>"physical"</span>,
<span class='kw'>LDmeasure</span><span class='kw'>=</span><span class='st'>"r"</span>, <span class='kw'>title</span><span class='kw'>=</span><span class='st'>"Pairwise LD"</span>, <span class='kw'>add.map</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='kw'>add.key</span><span class='kw'>=</span><span class='fl'>TRUE</span>,
<span class='kw'>geneMapLocation</span><span class='kw'>=</span><span class='fl'>0.15</span>, <span class='kw'>geneMapLabelX</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>geneMapLabelY</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>SNP.name</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>color</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>TRUE</span>,
<span class='kw'>name</span><span class='kw'>=</span><span class='st'>"ldheatmap"</span>, <span class='kw'>vp.name</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>pop</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>flip</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>text</span><span class='kw'>=</span><span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gdat</th>
      <td><p>SNP data: a data frame of <code>genotype</code> objects (a data structure from the <code>genetics</code> package), a <code>SnpMatrix</code> object (a data structure
from the <code>snpStats</code> BioConductor package), a square matrix of
pairwise linkage disequilibrium measurements or an object of
class <code>"LDheatmap"</code> (the returned object of this function).</p></td>
    </tr>
    <tr>
      <th>genetic.distances</th>
      <td><p>A numeric vector of map locations of the SNPs,
in the same order as SNPs listed in <code>gdat</code>,
in terms of genetic or physical distances.
Physical distances should be in bases, genetic distances should be in
centiMorgans (cM).
When <code>gdat</code> is <em>not</em> an object of class <code>LDheatmap</code>, the
default is a vector that represents equi-spaced markers, 1kb (1000 bases)
apart. When <code>gdat</code> <em>is</em> an object of class <code>LDheatmap</code>,
the <code>genetic.distances</code> argument is taken to be the
<code>genetic.distances</code> list item of <code>gdat</code>.</p></td>
    </tr>
    <tr>
      <th>distances</th>
      <td><p>A character string to specify whether the provided map locations
are in physical or genetic distances.
If <code>distances="physical"</code> (default), the text describing the total
length of the region will be &#8220;Physical Length:XXkb&#8221; where XX is the
length of the region in kilobases. If <code>distances="genetic"</code>, the
text will be &#8220;Genetic Map Length:YYcM&#8221; where YY is
the length of the region in centiMorgans.
If <code>gdat</code> is an object of class LDheatmap,
<code>distances</code> is taken from <code>gdat</code>.</p></td>
    </tr>
    <tr>
      <th>LDmeasure</th>
      <td><p>A character string specifying the measure of LD
- either allelic correlation \(r^2\) or Lewontin's
|D\('\)|; default = <code>"r"</code> for \(r^2\);
type <code>"D'"</code> for |D\('\)|.  This argument is ignored when the user has already
supplied calculated LD measurements through <code>gdat</code> (i.e., when <code>gdat</code>
is a matrix of pairwise LD measurements or an object of class <code>"LDheatmap"</code>).</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>A character string for the main title of the plot.
Default is &#8220;Pairwise LD&#8221;.</p></td>
    </tr>
    <tr>
      <th>add.map</th>
      <td><p>If <code>TRUE</code> (default) a diagonal line indicating
the physical or genetic map positions of the SNPs will be added to 
the plot, along with text indicating the total length of the 
genetic region.</p></td>
    </tr>
    <tr>
      <th>add.key</th>
      <td><p>If <code>TRUE</code> (default) the color legend is drawn.</p></td>
    </tr>
    <tr>
      <th>geneMapLocation</th>
      <td><p>A numeric value specifying the position of the line
parallel to the diagonal of the matrix; the larger the value, the
farther it lies from the matrix diagonal. Ignored when <code>add.map=FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>geneMapLabelX</th>
      <td><p>A numeric value specifying the x-coordinate
of the text indicating the total length of the genomic region
being considered. Ignored when <code>add.map=FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>geneMapLabelY</th>
      <td><p>A numeric value specifying the y-coordinate
of the text indicating the total length of the genomic region
being considered. Ignored when <code>add.map=FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>SNP.name</th>
      <td><p>A vector of character string(s) of SNP name(s) to
be labelled. Should match the names of SNPs in the provided object
<code>gdat</code>, otherwise nothing is done.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p>A range of colors to be used for drawing the heat map. Default</p></td>
    </tr>
    <tr>
      <th>newpage</th>
      <td><p>If <code>TRUE</code> (default), the heat map will be drawn on a new page.</p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>A character string specifying the name of the LDheatmap
graphical object (<code>grob</code>) to be produced.</p></td>
    </tr>
    <tr>
      <th>vp.name</th>
      <td><p>A character string specifying the name of the viewport
where the heat map is going to be drawn.</p></td>
    </tr>
    <tr>
      <th>pop</th>
      <td><p>If <code>TRUE</code>, the viewport where the heat map is drawn is
<code>pop</code>ped (i.e. removed) from the viewport tree after drawing. Default=<code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>flip</th>
      <td><p>If <code>TRUE</code>, the LDheatmap plot is flipped below a horizontal line, in the style of Haploview. Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>text</th>
      <td><p>If <code>TRUE</code>, the LD measurements are printed on each cell.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code>"LDheatmap"</code> which contains the following components:</p>
<dt>LDmatrix</dt><dd><p>The matrix of pairwise LD measurements plotted in the heat map.</p></dd>
<dt>LDheatmapGrob</dt><dd><p>A grid graphical object (grob) representing the produced heat map.</p></dd>
<dt>heatmapVP</dt><dd><p>The viewport in which the heat map is drawn. See <a href='http://www.rdocumentation.org/packages/grid/topics/viewport'>viewport</a>.</p></dd>
<dt>genetic.distances</dt><dd><p>The vector of the supplied physical or
genetic map locations, or the vector of equispaced marker distances
when no distance vector is supplied.</p></dd>
<dt>distances</dt><dd><p>A character string specifying whether the provided map
distances are physical or genetic.</p></dd>
<dt>color</dt><dd><p>The range of colors used for drawing the heat map.</p></dd>
The grob LDheatmapGrob has three grobs as its children (components).
They are listed below along with their own children and respectively represent
the color image with main title, genetic map and color key of the heat map:
 "heatMap" - "heatmap", "title";
"geneMap" - "diagonal", "segments",
"title", "symbols", "SNPnames"; and
"Key" - "colorKey", "title", "labels",
"ticks", "box".

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The input object <code>gdat</code> can be a data frame of <code>genotype</code> objects
(a data structure from the <span class="pkg">genetics</span> package), a <code>SnpMatrix</code> object (a
                                                                              data structure from the <span class="pkg">snpStats</span> package), or
any square matrix with values between 0 and 1
inclusive.
LD computation is much faster for <code>SnpMatrix</code> objects than for
<code>genotype</code> objects.
In the case of a matrix of LD values between 0 and 1,
the values above the diagonal will be plotted.
In the display of LD, SNPs appear in the order supplied by the
user as the horizontal and vertical coordinates are increased
and one moves along the off-diagonal line, from the bottom-left
to the top-right corner.  To achieve this, the conventions of
the <code>image()</code> function have been adopted, in which horizontal
coordinates correspond to the rows of the matrix and vertical coordinates
correspond to columns, and vertical coordinates are indexed in increasing
order from bottom to top.
For the argument <code>color</code>, an appropriate
color palette for quantitative data is recommended,
as outlined in the help page of
the <code><a href='http://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer'>brewer.pal</a>()</code> function of
the
<span class="pkg">RColorBrewer</span> package.
See the package vignette <code>LDheatmap</code> for more examples and details
of the implementation. Examples of adding ``tracks'' of genomic
annotation above a flipped heatmap are in the package vignette
<code>addTracks</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The produced heat map can be modified in two ways.
First, it is possible to edit <em>interactively</em> the grob components of the heat map,
by using the function <code><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></code>;
the function will not work if there is no
open graphical device showing the heat map.
Alternatively, the user can use the function
<code><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>editGrob</a></code> and work with
the grob <code>LDheatmapGrob</code> returned by <code>LDheatmap</code>.
See Examples for usage.
<code>LDheatmap()</code> uses <code><a href='http://www.rdocumentation.org/packages/grid/topics/Grid'>Grid</a></code>, which
does not respond to <code>par()</code> settings.
Hence modifying <code>par()</code> settings of <code>mfrow</code> and <code>mfcol</code>
will not work with <code>LDheatmap()</code>. The Examples section shows how to
display multiple heat maps on one plot without the use
of <code>par()</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Shin J-H, Blay S, McNeney B and Graham J (2006). LDheatmap:
An R Function for Graphical Display of Pairwise Linkage
Disequilibria Between Single Nucleotide Polymorphisms.
Journal of Statistical Software, <b>16</b> Code Snippet 3</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/genetics/topics/LD'>LD</a></code>, <code><a href='http://www.rdocumentation.org/packages/genetics/topics/genotype'>genotype</a></code>,
<code><a href='http://www.rdocumentation.org/packages/grid/topics/Grid'>Grid</a></code>, <code><a href='LDheatmap.highlight.html'>LDheatmap.highlight</a></code>,
<code><a href='LDheatmap.marks.html'>LDheatmap.marks</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Pass LDheatmap a SnpMatrix object</span>
<span class='fu'>set.seed</span>(<span class='fl'>1</span>)
<span class='co'>#make an example matrix of genotypes, coded as 0, 1 2 copies of an index allele</span>
<span class='no'>gdat</span><span class='kw'>&lt;-</span><span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='kw'>n</span><span class='kw'>=</span><span class='fl'>500</span>,<span class='kw'>size</span><span class='kw'>=</span><span class='fl'>2</span>,<span class='kw'>prob</span><span class='kw'>=</span><span class='fl'>.5</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>5</span>)
<span class='fu'>require</span>(<span class='no'>snpStats</span>)
<span class='no'>gdat</span><span class='kw'>&lt;-</span><span class='fu'>as</span>(<span class='no'>gdat</span>,<span class='st'>"SnpMatrix"</span>)</div><div class='output co'>#&gt; object has no names - using numeric order for row/column names</div><div class='input'><span class='fu'>LDheatmap</span>(<span class='no'>gdat</span>,<span class='kw'>genetic.distances</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>0</span>,<span class='fl'>1000</span>,<span class='fl'>3000</span>,<span class='fl'>4000</span>,<span class='fl'>10000</span>))</div><div class='img'><img src='LDheatmap-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Load the package's data set</span>
<span class='fu'>data</span>(<span class='no'>CEUData</span>)
<span class='co'>#Creates a data frame "CEUSNP" of genotype data and a vector "CEUDist"</span>
<span class='co'>#of physical locations of the SNPs</span>
<span class='co'># Produce a heat map in a grey color scheme</span>
<span class='no'>MyHeatmap</span> <span class='kw'>&lt;-</span> <span class='fu'>LDheatmap</span>(<span class='no'>CEUSNP</span>, <span class='kw'>genetic.distances</span> <span class='kw'>=</span> <span class='no'>CEUDist</span>,
                      <span class='kw'>color</span> <span class='kw'>=</span> <span class='fu'>grey.colors</span>(<span class='fl'>20</span>))</div><div class='input'><span class='co'># Same heatmap, flipped below a horizontal gene map -- for examples of</span>
<span class='co'># adding genomic annotation tracks to a flipped heatmap see</span>
<span class='co'># vignette("addTracks")</span>
<span class='co'># flippedHeatmap&lt;-LDheatmap(MyHeatmap,flip=TRUE)</span>
<span class='co'># Prompt the user before starting a new page of graphics output</span>
<span class='co'># and save the original prompt settings in old.prompt.</span>
<span class='no'>old.prompt</span> <span class='kw'>&lt;-</span> <span class='fu'>devAskNewPage</span>(<span class='kw'>ask</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># Highlight a certain LD block of interest:</span>
<span class='fu'><a href='LDheatmap.highlight.html'>LDheatmap.highlight</a></span>(<span class='no'>MyHeatmap</span>, <span class='kw'>i</span> <span class='kw'>=</span> <span class='fl'>3</span>, <span class='kw'>j</span> <span class='kw'>=</span> <span class='fl'>8</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"black"</span>,
<span class='kw'>fill</span> <span class='kw'>=</span> <span class='st'>"grey"</span>,<span class='kw'>flipOutline</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>crissCross</span><span class='kw'>=</span><span class='fl'>FALSE</span>)</div><div class='input'><span class='co'># Plot a symbol in the center of the pixel which represents LD between</span>
<span class='co'># the fourth and seventh SNPs:</span>
<span class='fu'><a href='LDheatmap.marks.html'>LDheatmap.marks</a></span>(<span class='no'>MyHeatmap</span>,  <span class='fl'>4</span>,  <span class='fl'>7</span>,  <span class='kw'>gp</span><span class='kw'>=</span><span class='kw pkg'>grid</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>cex</span><span class='kw'>=</span><span class='fl'>2</span>),  <span class='kw'>pch</span> <span class='kw'>=</span> <span class='st'>"*"</span>)</div><div class='img'><img src='LDheatmap-2.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#### Use an RGB pallete for the color scheme ####</span>
<span class='no'>rgb.palette</span> <span class='kw'>&lt;-</span> <span class='fu'>colorRampPalette</span>(<span class='fu'>rev</span>(<span class='fu'>c</span>(<span class='st'>"blue"</span>, <span class='st'>"orange"</span>, <span class='st'>"red"</span>)), <span class='kw'>space</span> <span class='kw'>=</span> <span class='st'>"rgb"</span>)
<span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>color</span><span class='kw'>=</span><span class='fu'>rgb.palette</span>(<span class='fl'>18</span>))</div><div class='img'><img src='LDheatmap-3.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#### Modify the plot by using 'grid.edit' function ####</span>
<span class='co'>#Draw a heat map where the SNPs "rs2283092" and "rs6979287" are labelled.</span>
<span class='fu'>require</span>(<span class='no'>grid</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: grid</span></div><div class='input'><span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>SNP.name</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"rs2283092"</span>, <span class='st'>"rs6979287"</span>))</div><div class='img'><img src='LDheatmap-4.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Find the names of the top-level graphical objects (grobs) on the current display</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/getNames'>getNames</a></span>()</div><div class='output co'>#&gt; [1] "ldheatmap"</div><div class='input'><span class='co'>#[1] "ldheatmap"</span>
<span class='co'># Find the names of the component grobs of "ldheatmap"</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.grob'>childNames</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.get'>grid.get</a></span>(<span class='st'>"ldheatmap"</span>))</div><div class='output co'>#&gt; [1] "heatMap" "geneMap" "Key"    </div><div class='input'><span class='co'>#[1] "heatMap" "geneMap" "Key"</span>
<span class='co'>#Find the names of the component grobs of heatMap</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.grob'>childNames</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.get'>grid.get</a></span>(<span class='st'>"heatMap"</span>))</div><div class='output co'>#&gt; [1] "heatmap" "title"  </div><div class='input'><span class='co'>#[1] "heatmap" "title"</span>
<span class='co'>#Find the names of the component grobs of geneMap</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.grob'>childNames</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.get'>grid.get</a></span>(<span class='st'>"geneMap"</span>))</div><div class='output co'>#&gt; [1] "diagonal" "segments" "title"    "symbols"  "SNPnames"</div><div class='input'><span class='co'>#[1] "diagonal" "segments" "title"    "symbols"  "SNPnames"</span>
<span class='co'>#Find the names of the component grobs of Key</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.grob'>childNames</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.get'>grid.get</a></span>(<span class='st'>"Key"</span>))</div><div class='output co'>#&gt; [1] "colorKey" "title"    "labels"   "ticks"    "box"     </div><div class='input'><span class='co'>#[1] "colorKey" "title"    "labels"   "ticks"    "box"</span>
<span class='co'>#Change the plotting symbols that identify SNPs rs2283092 and rs6979287</span>
<span class='co'>#on the plot to bullets</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='st'>"symbols"</span>, <span class='kw'>pch</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>gp</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>cex</span> <span class='kw'>=</span> <span class='fl'>1</span>))</div><div class='img'><img src='LDheatmap-5.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Change the color of the main title</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"ldheatmap"</span>, <span class='st'>"heatMap"</span>, <span class='st'>"title"</span>), <span class='kw'>gp</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"red"</span>))</div><div class='img'><img src='LDheatmap-6.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Change size of SNP labels</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"ldheatmap"</span>, <span class='st'>"geneMap"</span>,<span class='st'>"SNPnames"</span>), <span class='kw'>gp</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>cex</span><span class='kw'>=</span><span class='fl'>1.5</span>))</div><div class='img'><img src='LDheatmap-7.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Add a grid of white lines to the plot to separate pairwise LD measures</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"ldheatmap"</span>, <span class='st'>"heatMap"</span>, <span class='st'>"heatmap"</span>), <span class='kw'>gp</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"white"</span>,
                                                             <span class='kw'>lwd</span> <span class='kw'>=</span> <span class='fl'>2</span>))</div><div class='img'><img src='LDheatmap-8.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#### Modify a heat map using 'editGrob' function ####</span>
<span class='no'>MyHeatmap</span> <span class='kw'>&lt;-</span> <span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>color</span> <span class='kw'>=</span> <span class='fu'>grey.colors</span>(<span class='fl'>20</span>))</div><div class='img'><img src='LDheatmap-9.png' alt='' width='700' height='433' /></div><div class='input'><span class='no'>new.grob</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>editGrob</a></span>(<span class='no'>MyHeatmap</span>$<span class='no'>LDheatmapGrob</span>, <span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"geneMap"</span>, <span class='st'>"segments"</span>),
                    <span class='kw'>gp</span><span class='kw'>=</span><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>col</span><span class='kw'>=</span><span class='st'>"orange"</span>))
<span class='co'>##Clear the old graphics object from the display before drawing the modified heat map:</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.newpage'>grid.newpage</a></span>()
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.draw'>grid.draw</a></span>(<span class='no'>new.grob</span>)</div><div class='img'><img src='LDheatmap-10.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># now the colour of line segments connecting the SNP</span>
<span class='co'># positions to the LD heat map has been changed from black to orange.</span>
<span class='co'>#### Draw a resized heat map (in a 'blue-to-red' color scale ####</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.newpage'>grid.newpage</a></span>()
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>pushViewport</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewport'>viewport</a></span>(<span class='kw'>width</span><span class='kw'>=</span><span class='fl'>0.5</span>, <span class='kw'>height</span><span class='kw'>=</span><span class='fl'>0.5</span>))</div><div class='input'><span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>SNP.name</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"rs2283092"</span>, <span class='st'>"rs6979287"</span>), <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
         <span class='kw'>color</span><span class='kw'>=</span><span class='st'>"blueToRed"</span>)</div><div class='input'><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>popViewport</a></span>()</div><div class='img'><img src='LDheatmap-11.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#### Draw and modify two heat maps on one plot ####</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.newpage'>grid.newpage</a></span>()
<span class='co'>##Draw and the first heat map on the left half of the graphics device</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>pushViewport</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewport'>viewport</a></span>(<span class='kw'>x</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>width</span><span class='kw'>=</span><span class='fl'>0.5</span>, <span class='kw'>just</span><span class='kw'>=</span><span class='st'>"left"</span>))</div><div class='input'><span class='no'>LD1</span><span class='kw'>&lt;-</span><span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>color</span><span class='kw'>=</span><span class='fu'>grey.colors</span>(<span class='fl'>20</span>), <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
              <span class='kw'>title</span><span class='kw'>=</span><span class='st'>"Pairwise LD in grey.colors(20)"</span>,
              <span class='kw'>SNP.name</span><span class='kw'>=</span><span class='st'>"rs6979572"</span>, <span class='kw'>geneMapLabelX</span><span class='kw'>=</span><span class='fl'>0.6</span>,
              <span class='kw'>geneMapLabelY</span><span class='kw'>=</span><span class='fl'>0.4</span>, <span class='kw'>name</span><span class='kw'>=</span><span class='st'>"ld1"</span>)</div><div class='input'><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>upViewport</a></span>()</div><div class='input'><span class='co'>##Draw the second heat map on the right half of the graphics device</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>pushViewport</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewport'>viewport</a></span>(<span class='kw'>x</span><span class='kw'>=</span><span class='fl'>1</span>,<span class='kw'>width</span><span class='kw'>=</span><span class='fl'>0.5</span>,<span class='kw'>just</span><span class='kw'>=</span><span class='st'>"right"</span>))</div><div class='input'><span class='no'>LD2</span><span class='kw'>&lt;-</span><span class='fu'>LDheatmap</span>(<span class='no'>MyHeatmap</span>, <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>title</span><span class='kw'>=</span><span class='st'>"Pairwise LD in heat.colors(20)"</span>,
              <span class='kw'>SNP.name</span><span class='kw'>=</span><span class='st'>"rs6979572"</span>, <span class='kw'>geneMapLabelX</span><span class='kw'>=</span><span class='fl'>0.6</span>, <span class='kw'>geneMapLabelY</span><span class='kw'>=</span><span class='fl'>0.4</span>, <span class='kw'>name</span><span class='kw'>=</span><span class='st'>"ld2"</span>)</div><div class='input'><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/viewports'>upViewport</a></span>()</div><div class='img'><img src='LDheatmap-12.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>##Modify the text size of main title of the first heat map.</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"ld1"</span>, <span class='st'>"heatMap"</span>,<span class='st'>"title"</span>), <span class='kw'>gp</span><span class='kw'>=</span><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>cex</span><span class='kw'>=</span><span class='fl'>1.5</span>))</div><div class='img'><img src='LDheatmap-13.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>##Modify the text size and color of the SNP label of the second heat map.</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/grid.edit'>grid.edit</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gPath'>gPath</a></span>(<span class='st'>"ld2"</span>, <span class='st'>"geneMap"</span>,<span class='st'>"SNPnames"</span>), <span class='kw'>gp</span><span class='kw'>=</span><span class='fu'><a href='http://www.rdocumentation.org/packages/grid/topics/gpar'>gpar</a></span>(<span class='kw'>cex</span><span class='kw'>=</span><span class='fl'>1.5</span>, <span class='kw'>col</span><span class='kw'>=</span><span class='st'>"DarkRed"</span>))</div><div class='img'><img src='LDheatmap-14.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#### Draw a lattice-like plot with heat maps in panels ####</span>
<span class='co'># Load CHBJPTSNP and CHBJPTDist</span>
<span class='fu'>data</span>(<span class='no'>CHBJPTData</span>)
<span class='co'># Make a variable which indicates Chinese vs. Japanese</span>
<span class='no'>pop</span> <span class='kw'>&lt;-</span> <span class='fu'>factor</span>(<span class='fu'>c</span>(<span class='fu'>rep</span>(<span class='st'>"chinese"</span>,<span class='fl'>45</span>), <span class='fu'>rep</span>(<span class='st'>"japanese"</span>,<span class='fl'>45</span>)))
<span class='fu'>require</span>(<span class='no'>lattice</span>)
<span class='fu'><a href='http://www.rdocumentation.org/packages/lattice/topics/xyplot'>xyplot</a></span>(<span class='fl'>1</span>:<span class='fu'>nrow</span>(<span class='no'>CHBJPTSNP</span>) ~ <span class='fl'>1</span>:<span class='fu'>nrow</span>(<span class='no'>CHBJPTSNP</span>) <span class='kw'>|</span> <span class='no'>pop</span>,
      <span class='kw'>type</span><span class='kw'>=</span><span class='st'>"n"</span>, <span class='kw'>scales</span><span class='kw'>=</span><span class='fu'>list</span>(<span class='kw'>draw</span><span class='kw'>=</span><span class='fl'>FALSE</span>), <span class='kw'>xlab</span><span class='kw'>=</span><span class='st'>""</span>, <span class='kw'>ylab</span><span class='kw'>=</span><span class='st'>""</span>,
      <span class='kw'>panel</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='no'>subscripts</span>,<span class='no'>...</span>) {
        <span class='fu'>LDheatmap</span>(<span class='no'>CHBJPTSNP</span>[<span class='no'>subscripts</span>,], <span class='no'>CHBJPTDist</span>, <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>FALSE</span>) })</div><div class='img'><img src='LDheatmap-15.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>data</span>(<span class='no'>GIMAP5</span>)
<span class='fu'>require</span>(<span class='no'>lattice</span>)
<span class='no'>n</span><span class='kw'>&lt;-</span><span class='fu'>nrow</span>(<span class='no'>GIMAP5</span>$<span class='no'>snp.data</span>)
<span class='fu'><a href='http://www.rdocumentation.org/packages/lattice/topics/xyplot'>xyplot</a></span>(<span class='fl'>1</span>:<span class='no'>n</span> ~ <span class='fl'>1</span>:<span class='no'>n</span> <span class='kw'>|</span> <span class='no'>GIMAP5</span>$<span class='no'>subject.support</span>$<span class='no'>pop</span>,
      <span class='kw'>type</span><span class='kw'>=</span><span class='st'>"n"</span>, <span class='kw'>scales</span><span class='kw'>=</span><span class='fu'>list</span>(<span class='kw'>draw</span><span class='kw'>=</span><span class='fl'>FALSE</span>), <span class='kw'>xlab</span><span class='kw'>=</span><span class='st'>""</span>, <span class='kw'>ylab</span><span class='kw'>=</span><span class='st'>""</span>,
      <span class='kw'>panel</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='no'>subscripts</span>,<span class='no'>...</span>) {
        <span class='fu'>LDheatmap</span>(<span class='no'>GIMAP5</span>$<span class='no'>snp.data</span>[<span class='no'>subscripts</span>,],
                  <span class='no'>GIMAP5</span>$<span class='no'>snp.support</span>$<span class='no'>Position</span>, <span class='kw'>SNP.name</span><span class='kw'>=</span><span class='st'>"rs6598"</span>, <span class='kw'>newpage</span><span class='kw'>=</span><span class='fl'>FALSE</span>) })</div><div class='img'><img src='LDheatmap-16.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>#Reset the user's setting for prompting on the graphics output</span>
<span class='co'>#to the original value before running these example commands.</span>
<span class='fu'>devAskNewPage</span>(<span class='no'>old.prompt</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Ji-hyung Shin &lt;shin@sfu.ca&gt;, Sigal Blay &lt;sblay@sfu.ca&gt;,  Nicholas
Lewin-Koh &lt;nikko@hailmail.net&gt;, Brad McNeney &lt;mcneney@stat.sfu.ca&gt;, Jinko
Graham &lt;jgraham@cs.sfu.ca&gt;

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Ji-Hyung Shin, Sigal Blay, Brad McNeney, Jinko Graham.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

